### PR TITLE
feat: support docker for loader

### PR DESCRIPTION
**Pull request by aroundabout**
## Purpose of the PR

- support docker for hugegraph-loader

## Main Changes

1. Dockerfile for `loader`
2. an example docker-compose for loader

### usage

(add the usage to README and doc later)

current test image is  [dandelionivy/loader](https://hub.docker.com/repository/docker/dandelionivy/loader/general)

```bash
cd hugegraph-loader/docker/example/
docker-compose up -d 
docker exec -it loader bash
```

```bash
sh bin/hugegraph-loader.sh -g hugegraph -f example/file/struct.json -s example/file/schema.groovy -h graph -p 8080
```

It should be like:

```bash
user@LAPTOP-MRGBSP9S:~/incubator-hugegraph-toolchain/hugegraph-loader/docker/example$ sudo docker-compose up -d
Creating graph  ... done
Creating loader ... done
Creating hubble ... done
user@LAPTOP-MRGBSP9S:~/incubator-hugegraph-toolchain/hugegraph-loader/docker/example$ sudo docker ps
CONTAINER ID   IMAGE                 COMMAND                  CREATED              STATUS              PORTS                                         NAMES
59f431e622d0   hugegraph/hubble      "./bin/start-hubble.…"   About a minute ago   Up About a minute   0.0.0.0:8088->8088/tcp, :::8088->8088/tcp     hubble
55360d18f7d7   dandelionivy/loader   "/usr/bin/dumb-init …"   About a minute ago   Up About a minute                                                 loader
35b3feefc8bf   hugegraph/hugegraph   "/usr/bin/dumb-init …"   About a minute ago   Up About a minute   0.0.0.0:18081->8080/tcp, :::18081->8080/tcp   graph
user@LAPTOP-MRGBSP9S:~/incubator-hugegraph-toolchain/hugegraph-loader/docker/example$ sudo docker exec -it loader bash
root@55360d18f7d7:/loader# sh bin/hugegraph-loader.sh -g hugegraph -f example/file/struct.json -s example/file/schema.groovy -h graph -p 8080
...
vertices/edges loaded this time : 8/6
--------------------------------------------------
count metrics
    input read success            : 14                  
    input read failure            : 0                   
    vertex parse success          : 8                   
    vertex parse failure          : 0                   
    vertex insert success         : 8                   
    vertex insert failure         : 0                   
    edge parse success            : 6                   
    edge parse failure            : 0                   
    edge insert success           : 6                   
    edge insert failure           : 0                   
--------------------------------------------------
meter metrics
    total time                    : 0.199s              
    read time                     : 0.046s              
    load time                     : 0.153s              
    vertex load time              : 0.077s              
    vertex load rate(vertices/s)  : 103                 
    edge load time                : 0.112s              
    edge load rate(edges/s)       : 53                  
root@55360d18f7d7:/loader# 
```

Then we can see the result in `hubble`:

image

## Verifying these changes

- [ ] Trivial rework / code cleanup without any test coverage. (No Need)
- [ ] Already covered by existing tests, such as *(please modify tests here)*.
- [x] Need tests and can be verified as follows:
    - my test docker image is here [dandelionivy/loader](https://hub.docker.com/repository/docker/dandelionivy/loader/general)

## Does this PR potentially affect the following parts?

- [x]  Nope
- [ ]  Dependencies (add/update license info) 
- [ ]  Modify configurations
- [ ]  The public API
- [ ]  Other affects (typed here)

## Documentation Status

- [x]  `Doc - TODO` 
- [ ]  `Doc - Done` 
- [ ]  `Doc - No Need`

### CODE REVIEW

File: hugegraph-loader/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# limitations under the License.
+#
+
+FROM maven:3.9.0-eclipse-temurin-11 AS build
+
+COPY . /pkg
+WORKDIR /pkg
+
+RUN set -x \
+    && mvn install -pl hugegraph-client,hugegraph-loader -am -Dmaven.javadoc.skip=true -DskipTests -ntp
```
Comment (aroundabout) [inline hugegraph-loader/Dockerfile:24]: It seems that in the center maven repo, the client-1.0.0 does not have some methods like `istext()`. Hence, we should install the jar locally.

File: hugegraph-loader/Dockerfile
Comment (imbajin) [inline hugegraph-loader/Dockerfile:1]: we need also update the `README` in both root & loader level

tell users how to use the docker cmd for loader (and more **details** (like docker compose) in the README of loader level)

Comment (aroundabout) [reply hugegraph-loader/Dockerfile:1]: done

File: hugegraph-loader/Dockerfile (hunk 1 above)
Comment (imbajin) [reply hugegraph-loader/Dockerfile:24]: > It seems that in the center maven repo, the client-1.0.0 does not have some methods like `istext()`. Hence, we should install the jar locally.

seems we don't rely on fixed version? we use `$revision` in the project, so install client first is necessary (if we want build the latest image)

File: hugegraph-loader/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+       procps \
+       curl \
+       lsof \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+VOLUME /loader
+
+ENTRYPOINT ["/usr/bin/dumb-init", "--"]
+CMD [ "/bin/bash" ]
```
Comment (imbajin) [inline hugegraph-loader/Dockerfile:51]: maybe check other usage way for `dumb-init` ?

File: hugegraph-loader/README.md [hunk 3]
```diff
@@ -7,15 +7,121 @@
...
 - Multiple data sources, such as local file(path), HDFS file(path), MySQL
 - Various input formats, such as json, csv, and text with any delimiters.
 - Diverse options, with which users can manage the data loading intuitively.
 - Detecting schema from data automatically, reduce the complex work of schema management.
 - Advanced customized operations with groovy script, users can configure how to construct vertices and edges by themselves.
 
-## Building
+## 2. Usage for Docker(Recommand)
+
+- deploy `loader` with Docker
```
Comment (imbajin) [inline hugegraph-loader/README.md:20]:
```suggestion
- run `loader` with Docker
```

File: hugegraph-loader/README.md [hunk 4]
```diff
@@ -7,15 +7,121 @@
...
+```
+
+Use the command `docker-compose up -d` to deploy `loader` with `server` and `hubble`.
+
+### 2.2 Load data with docker container
+
+#### 2.2.1 enter the docker container
+
+```bash
+docker exec -it loader bash
```
Comment (imbajin) [inline hugegraph-loader/README.md:51]: remember update here when doc update https://github.com/apache/incubator-hugegraph-doc/pull/299#discussion_r1383507535

File: hugegraph-loader/README.md
Comment (imbajin) [inline hugegraph-loader/README.md:1]: consider also add a new section (run with docker) in root (README) https://github.com/apache/incubator-hugegraph-toolchain/blob/master/README.md#modules

and keep the `module` section clear
image

File: hugegraph-loader/Dockerfile (hunk 2 above)
Comment (aroundabout) [reply hugegraph-loader/Dockerfile:51]: It seems there is no special usage. We only need to add `dumb-init` in entrypoint. Maybe `tail -f /dev/null` is better than `/bin/bash`.

**Comment by liuxiaocs7** [comment]
Hi, @aroundabout, thanks for your outstanding efforts on this issue, i test locally and find some issues.

Firstly, i try to build the image locally but fails.

```
(base) kirin@kirinbiotech:~/lx/incubator-hugegraph-toolchain$ docker build -t hg-loader:1.0 hugegraph-loader/
```

image


Then, i use `loader image` from here: https://hub.docker.com/r/dandelionivy/loader/tags

image

```
kirin@kirinbiotech:~/lx/incubator-hugegraph-toolchain/hugegraph-loader/assembly/static$ docker exec -it loader bin/hugegraph-loader.sh -g hugegraph -f example/file/struct.json -s example/file/schema.groovy -p 8080 -h 192.168.110.158
```

image

but still fails to load data.

**Comment by aroundabout** [comment]
> Hi, @aroundabout, thanks for your outstanding efforts on this issue, i test locally and find some issues.
> 
> Firstly, i try to build the image locally but fails.
> 
> ```
> (base) kirin@kirinbiotech:~/lx/incubator-hugegraph-toolchain$ docker build -t hg-loader:1.0 hugegraph-loader/
> ```
> 
> image
> 
> Then, i use `loader image` from here: https://hub.docker.com/r/dandelionivy/loader/tags
> 
> image
> 
> ```
> kirin@kirinbiotech:~/lx/incubator-hugegraph-toolchain/hugegraph-loader/assembly/static$ docker exec -it loader bin/hugegraph-loader.sh -g hugegraph -f example/file/struct.json -s example/file/schema.groovy -p 8080 -h 192.168.110.158
> ```
> 
> image
> 
> but still fails to load data.

1. How to build locally
    try `docker build -f hugegraph-loader/Dockerfile hg-loader .`
2. The question is caused by server https://github.com/apache/incubator-hugegraph/issues/2346. The PR to fix it has been opened https://github.com/apache/incubator-hugegraph/pull/2347

**Comment by liuxiaocs7** [comment]
> How to build locally
try docker build -f hugegraph-loader/Dockerfile hg-loader .
Because loader depends on client and the client-1.0.0 in maven repo does not have some methods like 'istext()', so we should build client locally in dockerfile. Hence, the docker build context should be root path

Thanks for your details explain, i see.

```
docker build -f hugegraph-loader/Dockerfile hg-loader .
```

seems fails too? And does it need to be added to the readme?

image

**Comment by aroundabout** [comment]
> > How to build locally
> > try docker build -f hugegraph-loader/Dockerfile hg-loader .
> > Because loader depends on client and the client-1.0.0 in maven repo does not have some methods like 'istext()', so we should build client locally in dockerfile. Hence, the docker build context should be root path
> 
> Thanks for your details explain, i see.
> 
> ```
> docker build -f hugegraph-loader/Dockerfile hg-loader .
> ```
> 
> seems fails too? And does it need to be added to the readme?
> 
> image

Sorry, I forgot an argument. XD

`docker build -f hugegraph-loader/Dockerfile -t hg-loader .`

It can works now.
image

**Comment by liuxiaocs7** [comment]
> > > How to build locally
> > > try docker build -f hugegraph-loader/Dockerfile hg-loader .
> > > Because loader depends on client and the client-1.0.0 in maven repo does not have some methods like 'istext()', so we should build client locally in dockerfile. Hence, the docker build context should be root path
> > 
> > 
> > Thanks for your details explain, i see.
> > ```
> > docker build -f hugegraph-loader/Dockerfile hg-loader .
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > seems fails too? And does it need to be added to the readme?
> > image
> 
> Sorry, I forgot an argument. XD
> 
> `docker build -f hugegraph-loader/Dockerfile -t hg-loader .`
> 
> It can works now. image

image

works!

**Review by liuxiaocs7 — APPROVED** [review]
LGTM~

**Review by imbajin — APPROVED** [review]
THX

**Pull request merged**